`UnitControl`: fix controlled `unit` behavior

ciampo opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

In https://github.com/WordPress/gutenberg/pull/37311#issuecomment-1054591492, @ntsekouras flagged a bug where the `UnitControl` component wouldn't correctly update its unit if the update was coming via the `value` prop (as opposed as from a user interaction).
For reference, these two patterns are also referred to as "controlled" or "uncontrolled" components. `UnitControl` should technically already support this behaviour, but the bug report shows otherwise.

This PR introduces a fix around that, alongside a new Storybook example and a unit test that mimic closely the scenario encountered by Nik when flagging the buggy behavior.

This PR also introduces a new dev dependency, [`@testing-library/user-event`](https://testing-library.com/docs/ecosystem-user-event/), a recent addition to Testing Library for handling interaction in unit tests better.

**Potential follow-ups:**

- Follow-up on https://github.com/WordPress/gutenberg/pull/37311, rewriting the code in a way that doesn't use the deprecated `unit` prop
- Mark the `unit` prop as deprecated also in TypeScript types, and consider adding a soft deprecation warning too
- Rewrite Storybook examples to use controls (and convert to TypeScript)
- Rewrite unit tests to avoid using classnames anduse the `user-event` library (and convert to TypeScript)

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

- Run `UnitControl` unit tests, and make sure they pass
- Visit the "Controlled" Storybook example for `UnitControl` and make sure it works as expected (changing the unit in one control triggers that same unit change in the other control)

Comment out the fix added by this PR to the component:
- Run unit tests, the recently added test should fail
- Interact with the Storybook example, it should not work as expected

## Screenshots <!-- if applicable -->

Before:

https://user-images.githubusercontent.com/1083581/156259323-075f7d23-1e58-465e-8b53-78759f5ea428.mp4

After:

https://user-images.githubusercontent.com/1083581/156259333-c970c778-0768-4049-aec5-c59d87dd142c.mp4



## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->
- [ ] I've updated related schemas if appropriate. <!-- Reference: https://github.com/WordPress/gutenberg/tree/trunk/schemas -->

ciampo committed:
Add Storybook example

ciampo committed:
Fix UnitControl s controlled unit state

ciampo committed:
Add `@testing-library/user-event` to dev dependencies

ciampo committed:
Add unit test

ciampo committed:
CHANGELOG

andrewserong commented on package.json:
```diff
@@ -108,6 +108,7 @@
 		"@storybook/manager-webpack5": "6.4.9",
 		"@storybook/react": "6.4.9",
 		"@testing-library/jest-dom": "5.16.1",
+		"@testing-library/user-event": "13.5.0",
```
Linting is complaining that this line is not in alphabetical order 😀

amustaque97 commented:
Can we fix `static analysis` stage ` ✖ prefer-alphabetical-devDependencies - node: devDependencies - Your devDependencies are not in alphabetical order. Please move @testing-library/user-event after @testing-library/react.`

Changes LGTM 🚀

andrewserong reviewed (approved):
Nice fix @ciampo, thanks for the thoroughness with the addition of the storybook example and test coverage 👍 — that's all testing well for me, as is standard usage of the component in the block editor.

Aside from the linting issue in `package.json` this looks good to me, so I'll give it the ✅.

I like the addition of `user-event`, too, it's quite a bit easier to work with than `fireEvent` for more complex interactions. I'll just CC @gziolo, though, as he had feedback on another PR I believe that also was looking at adding in another devDependency (from memory, I think it was confidence checking how common it'll be to use the new things we add in as dependencies). In this case, since it looks like React Testing Library recommends using user-event most of the time in favour of `fireEvent` ([here in the docs](https://testing-library.com/docs/dom-testing-library/api-events/)), it sounds like it'll be worth it to add it in?

ciampo committed:
Fix package.json dependency alphabetical order

ciampo commented:
Thank you for the quick review!

I changed the order of the dependencies (somehow it wasn't flagged locally on my machine), and rebased to fix a conflict.

> In this case, since it looks like React Testing Library recommends using user-event most of the time in favour of fireEvent ([here in the docs](https://testing-library.com/docs/dom-testing-library/api-events/)), it sounds like it'll be worth it to add it in?

That's what I thought as well — `user-event` makes unit tests even easier to write _and_ read. I'll wait for @gziolo 's blessing before merging, though.

gziolo commented:
> I like the addition of `user-event`, too, it's quite a bit easier to work with than `fireEvent` for more complex interactions. I'll just CC @gziolo, though, as he had feedback on another PR I believe that also was looking at adding in another devDependency (from memory, I think it was confidence checking how common it'll be to use the new things we add in as dependencies). In this case, since it looks like React Testing Library recommends using user-event most of the time in favour of `fireEvent` ([here in the docs](https://testing-library.com/docs/dom-testing-library/api-events/)), it sounds like it'll be worth it to add it in?

There is one thing I don't like about the recommendations they included:

> Most projects have a few use cases for fireEvent, but the majority of the time you should probably use [@testing-library/user-event](https://testing-library.com/docs/ecosystem-user-event).

It too often leads to fragmentation in projects where you have old ways of writing tests, and suddenly there is a better way that impacts the development process because some people might consider it a best practice that everyone should follow. We still have tests written in Enzyme that were never refactored despite the fact that the project is no longer maintained by original creators. 

It looks like there is a growing need for having better testing tools so feel free to land this library as I see there is an agreement that the benefits are bigger than the potential maintenance cost. As far as I can tell, this library depends only on `@testing-library/dom` that is already in use.

For the future, I encourage you to make this type of proposal more transparent. A separate targeted PR with a refactored example test(s), documentation, a list of pros and cons included would give other contributors a chance to leave their feedback.

ciampo commented:
> For the future, I encourage you to make this type of proposal more transparent. A separate targeted PR with a refactored example test(s), documentation, a list of pros and cons included would give other contributors a chance to leave their feedback.

Hey @gziolo , you definitely make a good point here. I am also happy to remove the dependency, rewrite the tests with `fireEvent`, and open a separate PR which targets the new dependency separately. Let me know if you'd prefer this route instead.

ciampo committed:
Revert "Add `@testing-library/user-event` to dev dependencies"

This reverts commit 2c2200e827f35ac31b5b7bba39784cb329788ef3.

ciampo committed:
Rewrite unit tests without `user-event`

mirka commented on packages/components/src/unit-control/test/index.js:
```diff
@@ -25,6 +27,47 @@ const getUnitLabel = () =>
 const fireKeyDown = ( data ) =>
 	fireEvent.keyDown( document.activeElement || document.body, data );
 
+const ControlledSyncUnits = () => {
```
Non-blocking: Was this story only useful for testing this particular fix or do you think there is value in keeping it around? It's not immediately clear to me whether the main purpose of the story is to demonstrate a particular behavior/implementation, or to provide a debugging playground moving forward for the synced usage. My personal impression is that unit tests are sufficient and better for this kind of behavioral QA, unless you expect frequent manual testing of this "controlled" case in the near future.

ciampo replied:
You make a good point. This story is actually very targeted to test the buggy behaviour, and probably doesn't bring much value to Storybook users. I'll go ahead and remove it from this PR.

gziolo commented:
> > For the future, I encourage you to make this type of proposal more transparent. A separate targeted PR with a refactored example test(s), documentation, a list of pros and cons included would give other contributors a chance to leave their feedback.
> 
> Hey @gziolo , you definitely make a good point here. I am also happy to remove the dependency, rewrite the tests with `fireEvent`, and open a separate PR which targets the new dependency separately. Let me know if you'd prefer this route instead.

Both are fine. I don't have strong opinions on that at this point since I'm not opposed to the idea after a short investigation I did in regards to this small utilities library 😄

ciampo commented:
Update: I removed `@testing-library/user-event` from this PR and refactored the unit tests to use `fireEvent` instead.

I'll open a separate PR in the upcoming days about the introduction of the library to Gutenberg, highlighting the pros and showcasing a couple of examples.

I'll merge once CI is 🟢

mirka reviewed (commented):
Big +1 for adding `user-event` 😃

ciampo committed:
Remove temporary Storybook example

ntsekouras commented:
Thanks for the fix @ciampo!
